java/src/main/java/com/google/protobuf/ZeroCopyEncoder.java

```diff
@@ -0,0 +1,647 @@
+    if (!value.hasRemaining()) {
+      // Nothing to onDataEncoded.
+      return;
+    }
+
+    if (value.remaining() > SMALL_BUFFER) {
+      // It's a large buffer, just flush any previous data and onDataEncoded the buffer directly.
+      flush();
+
+      // Write a read-only view of this buffer. The new buffer has its own position and limit.
+      handler.onDataEncoded(value.asReadOnlyBuffer());
+
+      // Indicate that the buffer has been fully read.
+      value.position(value.limit());
+      return;
+    }
+
+    // Copy the value to the internal buffer, flushing as necessary.
+    while (value.hasRemaining()) {
+      int capacity = capacity();
+      if (capacity == 0) {
+        // The buffer is full, flush it and try again.
+        flush();
+        continue;
+      }
+
+      // Limit the value buffer by the number of bytes that can be written now.
+      int bytesToWrite = Math.min(value.remaining(), capacity);
+      int valueLimit = value.limit();
+      value.limit(value.position() + bytesToWrite);
+
+      getOrCreateBuffer().put(value);
+
+      // Restore the actual limit to the input value.
+      value.limit(valueLimit);
+    }
+  }
+
+  @Override
+  public void writeRawBytes(byte[] value, int offset, int length) throws IOException {
+    if (length == 0) {
+      return;
+    }
+
+    if (length > SMALL_BUFFER) {
+      // It's a large buffer, just flush any previous data and onDataEncoded the buffer directly.
+      flush();
+
+      // Write a read-only view of this buffer. The new buffer has its own position and limit.
+      handler.onDataEncoded(value, offset, length);
+      return;
+    }
+
+    while(offset < length) {
+      int capacity = capacity();
+      if (capacity == 0) {
+        // The buffer is full, flush it and try again.
+        flush();
+        continue;
+      }
+
+      // Determine the number of bytes that can be written before another flush is required.
+      int remaining = length - offset;
+      int bytesToWrite = Math.min(remaining, capacity);
+
+      getOrCreateBuffer().put(value, offset, bytesToWrite);
+      offset += bytesToWrite;
+    }
+  }
+
+  @Override
+  public void writeRawBytes(ByteString value, int offset, int length) throws IOException {
+    if (length == 0) {
+      return;
+    }
+
+    if (value instanceof LiteralByteString) {
+      LiteralByteString lbs = (LiteralByteString) value;
+      writeRawBytes(lbs.bytes, lbs.getOffsetIntoBytes(), lbs.size());
+    } else if (value instanceof UnsafeByteString) {
+      writeRawBytes(value.asReadOnlyByteBuffer());
+    } else {
+      for(ByteBuffer buf : value.asReadOnlyByteBufferList()) {
+        writeRawBytes(buf);
+      }
+    }
+  }
+
+  @Override
+  public void writeTag(int fieldNumber, int wireType) throws IOException {
+    writeRawVarint32(WireFormat.makeTag(fieldNumber, wireType));
+  }
+
+  @Override
+  public void writeRawVarint32(int value) throws IOException {
```

Comment:
Good idea.  Not sure if I'll do this as a part of this PR or not.  I think I'd like this PR to just be a comparison of zero-copy to the original ... without extra optimizations.